### ISSUE TITLE
Feature Request: Hiding the TODO keyword, with bullets replacing them visually

### ISSUE THREAD

**Issue by integral-dw** [integral-dw/org-superstar-mode#20]
>Not really a breakage of sorts but a feature request: a variable that will toggle hiding the actual keyword in place of the bullet replacement. I'm basically imagining org-mode turning into a convincing facimile of a Bullet Journal.
>
>This feature already is a great step in that direction but I'm still stuck looking at hoards of keywords in place of clean bullets and boxes.

_Originally posted by @ykhurshid in https://github.com/integral-dw/org-superstar-mode/issues/9#issuecomment-693511517_

**Comment by integral-dw** [comment]
Excellent point, @ykhurshid!  I do in fact have something along those lines planned in the future.  I have plans of making a minor mode that can turn many Org syntax elements into foldable menus, with TODO keywords being a natural application.  The reason this will require its own package is because it would add a lot of complexity that can be implemented orthogonally to Org Superstar.  The package (called `org-menu-mode`) would allow for hiding and unhiding based on cursor-position, meaning TODO keywords would disappear from view until you want to edit them.  The biggest hurdle in all of this is timing and efficiency: such a mode cannot afford to scale linearly with the number of menus in an org file (meaning I can't use overlays), but instead has to rely on deeper fontification, and sync seen parts of the buffer seamlessly after each command, meaning I have to deal with a lot of intricate `font-lock` fiddling, and forcing font-lock to re-inspect bits of the buffer that by no means *actually changed*.  The current alpha of org-menu-mode already "works", but there are plenty of subtle edge cases where it all falls apart, which makes debugging tremendously bothersome.

**TL;DR:** I actually have plans for supporting things like this (eventually maybe even independently of Org), but the code necessary is gonna be difficult to debug, and so I will need to wait for my schedule to be a bit more relaxed to properly work on it.

**Comment by ykhurshid** [comment]
Thanks for the response! I should have figured that the confusing crapshoot that is font-locking would complicate adding this functionality. That org-menu-mode package sounds very interesting though. Will definitely test it myself once the alpha is made public.

**Comment by integral-dw** [comment]
There actually is a public alpha version, so to say. Right now it's not particularly well-documented, but what documentation there is is still rough and likely full of typos. You can find the repo [here](https://github.com/integral-dw/org-menu-mode).  It's just race condition hell rn, so be warned.

**Comment by andersjohansson** [comment]
That looks cool @integral-dw!
In the meantime, for anyone seeing this and just wanting to turn your TODO keywords into symbols using prettify-symbols-mode, you can just do something like this:
```emacs-lisp
(defun my/org-enable-prettify ()
  (setq prettify-symbols-alist
        '(("TODO" . ?❢)
          ("DONE" . ?✔)
          ("STARTED" . ?➠)
          ("WAITING" . ?⌛)
          ("DISMISSED" . ?✘)))
  (prettify-symbols-mode))

(add-hook 'org-mode-hook #'my/org-enable-prettify)
```

**Comment by tbdcit** [comment]
Hi All, I just came to request the same thing. At the moment replacing the keywords with unicode symbols is a good compromise, however you still have the trailing bullet followed by the todo symbol. It would be great just to see the todo symbol. A more simple way of implementing this that could fit into org-superstar would be to have a setting in org-superstar that makes the  trailing bullet and the following space before the keyword invisible, i.e. 
```
* TODO
```
would become 
```
TODO
```
by applying `org-superstar--make-invisble` to the final `*` and the space. 

Is there any chance a feature like this could be made part of org-superstar?

**Comment by integral-dw** [comment]
I like the idea, @tbdcit!  I'll probably tinker with that on the volatile branch this weekend.

**Comment by integral-dw** [comment]
I pushed the new version to `volatile`, and it seems to work quite well.  The new feature will be included in version **1.5.0**.  You can then enable it by setting the variable `org-superstar-special-todo-items` to the symbol `hide`.  Until then you can emulate it by setting the characters in `org-superstar-todo-bullet-alist` to zero width spaces, like this:
```elisp
;; for Superstar 1.4.x
(setq org-superstar-todo-bullet-alist
  '(("TODO" . ?\N{ZERO WIDTH SPACE})
    ("DONE" . ?\N{ZERO WIDTH SPACE})))
(setq org-superstar-special-todo-items t)
```

**Comment by andersjohansson** [comment]
Nice! The current implementation missed todo states not in `org-superstar-todo-bullet-alist`. See #24

**Comment by integral-dw** [comment]
Ah, yes. That behavior is currently intended, the question is whether and how to give the user the option to specify a catch-all. 
I would argue that it would make sense to allow a special key value for `org-superstar-todo-bullet-alist`, such as "default" (a symbol, not a string).  Then, an entry such as `(default . "<generic bullet>")` would be orthogonal to existing options while not increasing the number of custom variables.  Sounds good?

**Comment by andersjohansson** [comment]
Ok! I see. But from my understanding, what bullets are defined in `org-superstar-todo-bullet-alist` shouldn’t matter when `org-superstar-special-todo-items` is set to `hide`, as they will be hidden? My use-case is that I use `prettify-symbols-mode` ([as explained here](#issuecomment-719539460)) for turning all my defined todo keywords into symbols, and then it is nice if `org-superstar` hides the bullet, and only shows the todo-keyword prettified into a symbol. But without modifying `org-superstar-todo-bullet-alist` some of my todo headings (all except `TODO` and `DONE`, which are defined in `org-superstar-todo-bullet-alist`) will have both the bullet (defined by level) and the keyword-symbol visible.

I cannot really understand the alternate use-case for using both `hide` and a detailed configuration of `org-superstar-todo-bullet-alist`. The possibility too choose which todo states should get the bullet hidden? But then customising the characters/bullets (cdr of entries) in `org-superstar-todo-bullet-alist` is unnecessary as only the states defined there play a role, and will be hidden. That seems kind of confusing to me.

Or am I misunderstanding something?

**Comment by integral-dw** [comment]
> I cannot really understand the alternate use-case for using both hide and a detailed configuration of org-superstar-todo-bullet-alist. 

Yes, `hide` and a custom `org-superstar-todo-bullet-alist` are normally mutually exclusive. Since `hide` is not in the master branch yet however (coming soon, I just wanna test one or two more things), one can use the alist to emulate hiding. That is the trick I mentioned above.
> But without modifying org-superstar-todo-bullet-alist some of my todo headings (all except TODO and DONE, which are defined in org-superstar-todo-bullet-alist) will have both the bullet (defined by level) and the keyword-symbol visible.

That's what I fixed, you can have a default entry now which serves as a fallback.  See above.

>Or am I misunderstanding something?

The patch just allows for a temporary hack to work, but also opens up a feature completely unrelated to hiding (having a generic TODO bullet), which is why I implemented it.

**Comment by integral-dw** [comment]
Welp, turns out I can't read. Your PR was perfect for fixing the bug in the volatile branch. What I thought you meant was the very same bug the hack had.  Talk about tunnel vision.  Well, I also introduced a new bug now, which will be fixed in a hot minute.  Pardon the confusion, guess I sucked at reading comprehension the past couple days.

**Comment by integral-dw** [comment]
And **v1.5.0** is now released!

**Comment by andersjohansson** [comment]
Ok, thanks! Looks great! 😁
